Add support for arbitrary beat divisors

bdach opened the pull request:
...Or, well, up to 1/64 I guess. I added that limit only to keep some semblance of sanity. I'd hope nobody seriously ever needs higher.

https://user-images.githubusercontent.com/20418176/155896470-eb266b3f-a6e8-428b-a57c-94fa0c0b49be.mp4

The [feedback](https://github.com/ppy/osu/issues/2316#issuecomment-1038774385) from my initial attempt has been addressed, I think.

Stable obviously does not (and will not) support the fully custom presets aside from what it already can support but it degrades "gracefully" (i.e. doesn't die and instead chooses smallest closer available divisor).

Diff's a bit large again but a substantial portion of it is tests.

* Closes #2316
* Supersedes / closes #2403

bdach committed:
Add simple carousel divisor type selector

bdach committed:
Add flow for changing set of valid divisors between presets

bdach committed:
Add flow for specifying entirely custom snaps

bdach committed:
Add test coverage for divisor behaviour

bdach committed:
Fix test failures

Walavouchey commented:
The small wall of text in the popup can be simplified to "All divisors will be added to the list of presets.", if it even is necessary

bdach commented:
that's not even an accurate simplification. the word "applicable" is key there. there is 0 sense in adding a 1/5 snap when you're choosing 1/7, for instance, and it would look way too crowded. the way that works is that it takes the divisor that the user inputs and finds all smaller snaps that are divisor of the snap being provided. so for instance choosing 1/28 will add 1/1, 1/2, 1/4 and 1/7 to presets.

Walavouchey commented:
The key word is "divisor". Of course 1/5 won't be added for 1/7, because 5 isn't a divisor of 7.

bdach commented:
did you accidentally a word there perhaps? i'm seeing "all divisors". did you perhaps mean "all *smaller* divisors"? because i can't see how the former version isn't just flat out wrong.

Walavouchey commented:
Revising my suggestion to "All divisors of this beat snap will be added to the list of presets", as per [small discussion on discord](https://discord.com/channels/188630481301012481/188630652340404224/947578872336105554)

bdach committed:
Fix test failure due to attempting to set non-present divisor

With the latest changes permitting fully custom beat snapping, the 1/3
snap divisor isn't immediately available in editor, requiring a switch
to "triplets" mode first.

peppy committed:
Automatically focus divisor textbox and hide popover after successful change

peppy committed:
Remove unused using statement

peppy committed:
Reword popup text to read better (or more vaguely)

Removed some words but also don't mention "smaller" because it's...
musically incorrect and also functionally incorrect – entering 1/[8]
will result in 1/16 also being populated for instance.

peppy commented:
I've reworded the popup text based on the discussion here (see 2be40f36f77c81501d5530ff136908111ce3ddb8), but decided on different wording. Will leave it up to your discretion on whether it needs further changes. I think making it read less specifically is fine and will let us tweak it going forward. ie., if the user enters `1/5` we may want to populated all the way up to `1/15` for them, if that's seen as beneficial.

peppy reviewed (approved):
Not going to over-think this one. Structurally and functionally feels good.

bdach commented:
I can get behind the new popup text for sure.
